docs/proposals/20-ha-control-planes.md

```diff
@@ -0,0 +1,1259 @@
+      - labelSelector:
+          matchLabels:
+            <labels>
+        topologyKey: "kubernetes.io/hostname"
+```
+
+### Multi-Zone (#replicas == #zones)
+If the replica count is equal to the number of available zones, then we can enforce the zone spread during scheduling.
+
+```yaml
+spec:
+  affinity:
+    podAntiAffinity:
+      requiredDuringSchedulingIgnoredDuringExecution:
+      - labelSelector:
+          matchLabels:
+            <labels>
+      topologyKey: "topology.kubernetes.io/zone"
+```
+
+### Multi-Zone (#replicas > #zones)
+Enforcing a zone spread for components with a replica count higher than the total amount of zones is not possible. In this case we plan to rather use the following [Pod Topology Spread Constraints](https://kubernetes.io/docs/concepts/workloads/pods/pod-topology-spread-constraints/) which allows a distribution over zones and nodes. The `maxSkew` value determines how big a imbalance of the pod distribution can be and thus it allows to schedule replicas with a count beyond the number of availability zones (e.g. [Kube-Apiserver](#kube-apiserver)).
+
+> **NOTE:**
+>
+> During testing we found a few inconsistencies and some quirks (more [information](#topology-spread-constraints-evaluation-and-findings)) which is why we rely on Topology Spread Constraints (TSC) only for this case.
+> In addition to circumvent issue [kubernetes/kubernetes#98215](https://github.com/kubernetes/kubernetes/issues/98215), Gardener is supposed to add a `gardener.cloud/rolloutVersion` label and incrementing the version every time the `.spec` of the component is changed (see [workaround](https://github.com/kubernetes/kubernetes/issues/98215#issuecomment-766146323)).
+
+```yaml
+spec:
+  topologySpreadConstraints:
+  - maxSkew: 2
+    topologyKey: topology.kubernetes.io/zone
+    whenUnsatisfiable: DoNotSchedule
+    labelSelector:
+      matchLabels:
+        <labels>
+        gardener.cloud/rolloutVersion: <version>
+  - maxSkew: 1
+    topologyKey: kubernetes.io/hostname
+    whenUnsatisfiable: DoNotSchedule
+    labelSelector:
+      matchLabels:
+        <labels>
+        gardener.cloud/rolloutVersion: <version>
+```
+
+## Disruptions and zero downtime maintenance
+
+A secondary effect of provisioning affected seed system and shoot control plane components in an HA fashion is the support for zero downtime maintenance, i.e. a certain amount of replicas always serves requests while an update is being rolled out.
+Therefore, proper [PodDisruptionBudgets](https://kubernetes.io/docs/concepts/workloads/pods/disruptions/) are required. With this GEP it's planned to set `spec.maxUnavailable: 1` for every involved and further mentioned component.
+
+## Seed System Components
+
+The following seed system components already run or are planned[*] to be configured with a minimum of two replicas:
+
+- Etcd-Druid* (active-passive)
+- Gardenlet* (active-passive)
+- Istio Ingress Gateway* (active-active)
```

Comment:
Not relevant in the scope of this GEP, but can we run this one here with at least one replica per zone and topology-aware routing passing on the requests to the API servers in the same zone?

What can we do for non-HA and single-zone HA shoots? Here, it would be ideal to receive the traffic already in the "right" zone.

Reply:
Good question! Looping in @ScheererJ and @DockToFuture. Do you have an idea or already tried out a similar scenario?

Reply:
Uh, that's then expensive...